File: api/parsers/markdown_parser.py
"""
Markdown Parser for Playbook Forge

Converts structured markdown playbooks into a node/edge graph format compatible with React Flow.

Parsing Rules:
- H1/H2 headers become phase nodes
- Numbered lists become sequential step nodes
- Bullet points with 'if/when/else' become decision nodes with branches
- Code blocks become 'execute' nodes with code content
"""

import re
from typing import List, Dict, Any, Tuple, Optional
from api.models import PlaybookNode, PlaybookEdge, PlaybookGraph


class MarkdownParser:
    """Parser for converting markdown playbooks to flowchart graphs."""

    def __init__(self):
        """Initialize the markdown parser."""
        self.nodes: List[PlaybookNode] = []
        self.edges: List[PlaybookEdge] = []
        self.node_counter = 0
        self.edge_counter = 0
        self.last_node_id: Optional[str] = None
        self.phase_stack: List[str] = []  # Track current phase hierarchy

    def parse(self, content: str) -> PlaybookGraph:
        """
        Parse markdown content into a flowchart graph.

        Args:
            content: Markdown string to parse

        Returns:
            PlaybookGraph with nodes and edges
        """
        # Reset state for fresh parse
        self.nodes = []
        self.edges = []
        self.node_counter = 0
        self.edge_counter = 0
        self.last_node_id = None
        self.phase_stack = []

        lines = content.split('\n')
        i = 0

        while i < len(lines):
            line = lines[i]
            stripped = line.strip()

            # Skip empty lines
            if not stripped:
                i += 1
                continue

            # Parse headers (H1/H2) as phase nodes
            if stripped.startswith('#'):
                self._parse_header(stripped)
                i += 1
                continue

            # Parse numbered lists as sequential steps
            if re.match(r'^\d+\.', stripped):
                self._parse_numbered_list_item(stripped)
                i += 1
                continue

            # Parse bullet points (check for decision keywords)
            if stripped.startswith('-') or stripped.startswith('*'):
                i = self._parse_bullet_point(lines, i)
                continue

            # Parse code blocks as execute nodes
            if stripped.startswith('```'):
                i = self._parse_code_block(lines, i)
                continue

            i += 1

        return PlaybookGraph(nodes=self.nodes, edges=self.edges)

    def _parse_header(self, line: str) -> None:
        """
        Parse markdown header into a phase node.

        Args:
            line: Header line (e.g., "# Phase 1" or "## Setup")
        """
        # Count hashes to determine header level
        level = 0
        for char in line:
            if char == '#':
                level += 1
            else:
                break

        label = line.lstrip('#').strip()
        node_id = self._create_node_id()

        # Create phase node
        node = PlaybookNode(
            id=node_id,
            label=label,
            type="phase",
            metadata={"level": level, "header_type": f"h{level}"}
        )
        self.nodes.append(node)

        # Connect to previous node if exists
        if self.last_node_id:
            self._create_edge(self.last_node_id, node_id)

        self.last_node_id = node_id

        # Update phase stack
        if level == 1:
            self.phase_stack = [node_id]
        elif level == 2 and self.phase_stack:
            self.phase_stack = [self.phase_stack[0], node_id]

    def _parse_numbered_list_item(self, line: str) -> None:
        """
        Parse numbered list item into a sequential step node.

        Args:
            line: Numbered list line (e.g., "1. Install dependencies")
        """
        # Extract the text after the number
        match = re.match(r'^\d+\.\s*(.+)$', line)
        if not match:
            return

        label = match.group(1).strip()
        node_id = self._create_node_id()

        # Create step node
        node = PlaybookNode(
            id=node_id,
            label=label,
            type="step",
            metadata={"step_type": "sequential"}
        )
        self.nodes.append(node)

        # Connect to previous node
        if self.last_node_id:
            self._create_edge(self.last_node_id, node_id)

        self.last_node_id = node_id

    def _parse_bullet_point(self, lines: List[str], start_idx: int) -> int:
        """
        Parse bullet point, detecting decision nodes with if/when/else keywords.

        Args:
            lines: All lines in the document
            start_idx: Current line index

        Returns:
            Next line index to process
        """
        line = lines[start_idx].strip()

        # Extract bullet content
        if line.startswith('-'):
            content = line[1:].strip()
        elif line.startswith('*'):
            content = line[1:].strip()
        else:
            return start_idx + 1

        # Check for decision keywords
        decision_keywords = ['if ', 'when ', 'else', 'otherwise', 'or if', 'elif']
        is_decision = any(keyword in content.lower() for keyword in decision_keywords)

        if is_decision:
            return self._parse_decision_node(lines, start_idx, content)
        else:
            # Regular bullet point - treat as step
            node_id = self._create_node_id()
            node = PlaybookNode(
                id=node_id,
                label=content,
                type="step",
                metadata={"step_type": "bullet"}
            )
            self.nodes.append(node)

            if self.last_node_id:
                self._create_edge(self.last_node_id, node_id)

            self.last_node_id = node_id
            return start_idx + 1

    def _parse_decision_node(self, lines: List[str], start_idx: int, content: str) -> int:
        """
        Parse a decision node with branches.

        Args:
            lines: All lines in the document
            start_idx: Current line index
            content: Content of the decision line

        Returns:
            Next line index to process
        """
        # Create decision node
        decision_id = self._create_node_id()

        # Extract condition text
        condition = self._extract_condition(content)

        node = PlaybookNode(
            id=decision_id,
            label=condition,
            type="decision",
            metadata={"condition": content}
        )
        self.nodes.append(node)

        # Connect to previous node
        if self.last_node_id:
            self._create_edge(self.last_node_id, decision_id)

        # Look ahead for nested items (branches)
        idx = start_idx + 1
        branches: List[Tuple[str, str]] = []  # (label, content) pairs

        while idx < len(lines):
            next_line = lines[idx]
            stripped = next_line.strip()

            # Check if this is an indented item (branch)
            if next_line.startswith('  -') or next_line.startswith('  *') or next_line.startswith('    -') or next_line.startswith('    *'):
                branch_content = stripped[1:].strip()

                # Determine branch label
                if 'else' in content.lower() or 'otherwise' in content.lower():
                    branch_label = 'no'
                elif branches:
                    branch_label = 'no'
                else:
                    branch_label = 'yes'

                branches.append((branch_label, branch_content))
                idx += 1
            else:
                break

        # Create branch nodes
        merge_node_id: Optional[str] = None

        for branch_label, branch_content in branches:
            branch_node_id = self._create_node_id()
            branch_node = PlaybookNode(
                id=branch_node_id,
                label=branch_content,
                type="step",
                metadata={"branch": branch_label}
            )
            self.nodes.append(branch_node)

            # Connect decision to branch
            self._create_edge(decision_id, branch_node_id, label=branch_label)

            # Create merge point if multiple branches
            if not merge_node_id and len(branches) > 1:
                merge_node_id = self._create_node_id()
                merge_node = PlaybookNode(
                    id=merge_node_id,
                    label="Continue",
                    type="merge",
                    metadata={"merge_type": "decision"}
                )
                self.nodes.append(merge_node)

            # Connect branch to merge
            if merge_node_id:
                self._create_edge(branch_node_id, merge_node_id)

        # Update last_node_id to merge point or decision node
        self.last_node_id = merge_node_id if merge_node_id else decision_id

        return idx

    def _parse_code_block(self, lines: List[str], start_idx: int) -> int:
        """
        Parse code block into an execute node.

        Args:
            lines: All lines in the document
            start_idx: Current line index (should be opening ```)

        Returns:
            Next line index to process
        """
        opening_line = lines[start_idx].strip()
        language = opening_line[3:].strip() if len(opening_line) > 3 else ""

        # Collect code content
        code_lines = []
        idx = start_idx + 1

        while idx < len(lines):
            if lines[idx].strip().startswith('```'):
                # Found closing backticks
                break
            code_lines.append(lines[idx])
            idx += 1

        code_content = '\n'.join(code_lines)

        # Create execute node
        node_id = self._create_node_id()
        node = PlaybookNode(
            id=node_id,
            label=f"Execute {language}" if language else "Execute code",
            type="execute",
            metadata={
                "code": code_content,
                "language": language
            }
        )
        self.nodes.append(node)

        # Connect to previous node
        if self.last_node_id:
            self._create_edge(self.last_node_id, node_id)

        self.last_node_id = node_id

        # Return index after closing backticks
        return idx + 1

    def _extract_condition(self, content: str) -> str:
        """
        Extract clean condition text from decision content.

        Args:
            content: Full content of decision line

        Returns:
            Cleaned condition text
        """
        # Remove common prefixes
        for prefix in ['if ', 'when ', 'else ', 'otherwise ', 'or if ', 'elif ']:
            if content.lower().startswith(prefix):
                content = content[len(prefix):].strip()
                break

        # Capitalize first letter
        if content:
            content = content[0].upper() + content[1:]

        # Add question mark if not present
        if content and not content.endswith('?'):
            content += '?'

        return content

    def _create_node_id(self) -> str:
        """
        Generate unique node ID.

        Returns:
            Node ID string
        """
        node_id = f"node_{self.node_counter}"
        self.node_counter += 1
        return node_id

    def _create_edge(self, source: str, target: str, label: Optional[str] = None) -> None:
        """
        Create an edge between two nodes.

        Args:
            source: Source node ID
            target: Target node ID
            label: Optional edge label
        """
        edge_id = f"edge_{self.edge_counter}"
        self.edge_counter += 1

        edge = PlaybookEdge(
            id=edge_id,
            source=source,
            target=target,
            label=label
        )
        self.edges.append(edge)

File: web/src/parsers/markdownParser.ts
/**
 * Markdown Parser for Playbook Forge (Client-Side)
 *
 * TypeScript port of api/parsers/markdown_parser.py
 *
 * Converts structured markdown playbooks into a node/edge graph format
 * compatible with React Flow.
 *
 * Parsing Rules:
 * - H1/H2 headers become phase nodes
 * - Numbered lists become sequential step nodes
 * - Bullet points with 'if/when/else' become decision nodes with branches
 * - Code blocks become 'execute' nodes with code content
 * - Automatic edges between sequential elements
 */

import {
  PlaybookNode,
  PlaybookEdge,
  PlaybookGraph,
  PlaybookMetadata,
  ParseResult,
  NodeType,
} from '../types';

class MarkdownParser {
  private nodes: PlaybookNode[] = [];
  private edges: PlaybookEdge[] = [];
  private nodeCounter = 0;
  private edgeCounter = 0;
  private lastNodeId: string | null = null;
  private phaseStack: string[] = [];

  /**
   * Parse markdown content into a flowchart graph.
   */
  parse(content: string): PlaybookGraph {
    // Reset state for fresh parse
    this.nodes = [];
    this.edges = [];
    this.nodeCounter = 0;
    this.edgeCounter = 0;
    this.lastNodeId = null;
    this.phaseStack = [];

    const lines = content.split('\n');
    let i = 0;

    while (i < lines.length) {
      const line = lines[i];
      const stripped = line.trim();

      // Skip empty lines
      if (!stripped) {
        i++;
        continue;
      }

      // Parse headers (H1/H2) as phase nodes
      if (stripped.startsWith('#')) {
        this.parseHeader(stripped);
        i++;
        continue;
      }

      // Parse numbered lists as sequential steps
      if (/^\d+\./.test(stripped)) {
        this.parseNumberedListItem(stripped);
        i++;
        continue;
      }

      // Parse bullet points (check for decision keywords)
      if (stripped.startsWith('-') || stripped.startsWith('*')) {
        i = this.parseBulletPoint(lines, i);
        continue;
      }

      // Parse code blocks as execute nodes
      if (stripped.startsWith('```')) {
        i = this.parseCodeBlock(lines, i);
        continue;
      }

      i++;
    }

    return { nodes: this.nodes, edges: this.edges };
  }

  /**
   * Parse markdown header into a phase node.
   */
  private parseHeader(line: string): void {
    let level = 0;
    for (const char of line) {
      if (char === '#') {
        level++;
      } else {
        break;
      }
    }

    const label = line.replace(/^#+\s*/, '').trim();
    // Skip emoji-only or empty labels
    if (!label) return;

    const nodeId = this.createNodeId();

    const node: PlaybookNode = {
      id: nodeId,
      label,
      type: 'phase',
      metadata: { level, header_type: `h${level}` },
    };
    this.nodes.push(node);

    // Connect to previous node if exists
    if (this.lastNodeId) {
      this.createEdge(this.lastNodeId, nodeId);
    }

    this.lastNodeId = nodeId;

    // Update phase stack
    if (level === 1) {
      this.phaseStack = [nodeId];
    } else if (level === 2 && this.phaseStack.length > 0) {
      this.phaseStack = [this.phaseStack[0], nodeId];
    }
  }

  /**
   * Parse numbered list item into a sequential step node.
   */
  private parseNumberedListItem(line: string): void {
    const match = line.match(/^\d+\.\s*(.+)$/);
    if (!match) return;

    const label = match[1].trim();
    const nodeId = this.createNodeId();

    const node: PlaybookNode = {
      id: nodeId,
      label,
      type: 'step',
      metadata: { step_type: 'sequential' },
    };
    this.nodes.push(node);

    if (this.lastNodeId) {
      this.createEdge(this.lastNodeId, nodeId);
    }

    this.lastNodeId = nodeId;
  }

  /**
   * Parse bullet point, detecting decision nodes with if/when/else keywords.
   */
  private parseBulletPoint(lines: string[], startIdx: number): number {
    const line = lines[startIdx].trim();

    let content: string;
    if (line.startsWith('-')) {
      content = line.slice(1).trim();
    } else if (line.startsWith('*')) {
      content = line.slice(1).trim();
    } else {
      return startIdx + 1;
    }

    // Check for decision keywords
    const decisionKeywords = ['if ', 'when ', 'else', 'otherwise', 'or if', 'elif'];
    const isDecision = decisionKeywords.some((kw) =>
      content.toLowerCase().includes(kw)
    );

    if (isDecision) {
      return this.parseDecisionNode(lines, startIdx, content);
    } else {
      // Regular bullet point - treat as step
      const nodeId = this.createNodeId();
      const node: PlaybookNode = {
        id: nodeId,
        label: content,
        type: 'step',
        metadata: { step_type: 'bullet' },
      };
      this.nodes.push(node);

      if (this.lastNodeId) {
        this.createEdge(this.lastNodeId, nodeId);
      }

      this.lastNodeId = nodeId;
      return startIdx + 1;
    }
  }

  /**
   * Parse a decision node with branches.
   */
  private parseDecisionNode(
    lines: string[],
    startIdx: number,
    content: string
  ): number {
    const decisionId = this.createNodeId();
    const condition = this.extractCondition(content);

    const node: PlaybookNode = {
      id: decisionId,
      label: condition,
      type: 'decision',
      metadata: { condition: content },
    };
    this.nodes.push(node);

    if (this.lastNodeId) {
      this.createEdge(this.lastNodeId, decisionId);
    }

    // Look ahead for nested items (branches)
    let idx = startIdx + 1;
    const branches: Array<{ label: string; content: string }> = [];

    while (idx < lines.length) {
      const nextLine = lines[idx];
      const stripped = nextLine.trim();

      // Check if this is an indented item (branch)
      if (
        nextLine.startsWith('  -') ||
        nextLine.startsWith('  *') ||
        nextLine.startsWith('    -') ||
        nextLine.startsWith('    *')
      ) {
        const branchContent = stripped.slice(1).trim();

        // Determine branch label
        let branchLabel: string;
        if (
          content.toLowerCase().includes('else') ||
          content.toLowerCase().includes('otherwise')
        ) {
          branchLabel = 'no';
        } else if (branches.length > 0) {
          branchLabel = 'no';
        } else {
          branchLabel = 'yes';
        }

        branches.push({ label: branchLabel, content: branchContent });
        idx++;
      } else {
        break;
      }
    }

    // Create branch nodes
    let mergeNodeId: string | null = null;

    for (const branch of branches) {
      const branchNodeId = this.createNodeId();
      const branchNode: PlaybookNode = {
        id: branchNodeId,
        label: branch.content,
        type: 'step',
        metadata: { branch: branch.label },
      };
      this.nodes.push(branchNode);

      // Connect decision to branch
      this.createEdge(decisionId, branchNodeId, branch.label);

      // Create merge point if multiple branches
      if (!mergeNodeId && branches.length > 1) {
        mergeNodeId = this.createNodeId();
        const mergeNode: PlaybookNode = {
          id: mergeNodeId,
          label: 'Continue',
          type: 'step' as NodeType, // 'merge' maps to step for rendering
          metadata: { merge_type: 'decision' },
        };
        this.nodes.push(mergeNode);
      }

      // Connect branch to merge
      if (mergeNodeId) {
        this.createEdge(branchNodeId, mergeNodeId);
      }
    }

    // Update lastNodeId to merge point or decision node
    this.lastNodeId = mergeNodeId ?? decisionId;

    return idx;
  }

  /**
   * Parse code block into an execute node.
   */
  private parseCodeBlock(lines: string[], startIdx: number): number {
    const openingLine = lines[startIdx].trim();
    const language =
      openingLine.length > 3 ? openingLine.slice(3).trim() : '';

    const codeLines: string[] = [];
    let idx = startIdx + 1;

    while (idx < lines.length) {
      if (lines[idx].trim().startsWith('```')) {
        break;
      }
      codeLines.push(lines[idx]);
      idx++;
    }

    const codeContent = codeLines.join('\n');
    const nodeId = this.createNodeId();

    const node: PlaybookNode = {
      id: nodeId,
      label: language ? `Execute ${language}` : 'Execute code',
      type: 'execute',
      metadata: {
        code: codeContent,
        language,
      },
    };
    this.nodes.push(node);

    if (this.lastNodeId) {
      this.createEdge(this.lastNodeId, nodeId);
    }

    this.lastNodeId = nodeId;

    // Return index after closing backticks
    return idx + 1;
  }

  /**
   * Extract clean condition text from decision content.
   */
  private extractCondition(content: string): string {
    let cleaned = content;
    const prefixes = [
      'if ',
      'when ',
      'else ',
      'otherwise ',
      'or if ',
      'elif ',
    ];

    for (const prefix of prefixes) {
      if (cleaned.toLowerCase().startsWith(prefix)) {
        cleaned = cleaned.slice(prefix.length).trim();
        break;
      }
    }

    // Capitalize first letter
    if (cleaned) {
      cleaned = cleaned[0].toUpperCase() + cleaned.slice(1);
    }

    // Add question mark if not present
    if (cleaned && !cleaned.endsWith('?')) {
      cleaned += '?';
    }

    return cleaned;
  }

  /**
   * Generate unique node ID.
   */
  private createNodeId(): string {
    const id = `node_${this.nodeCounter}`;
    this.nodeCounter++;
    return id;
  }

  /**
   * Create an edge between two nodes.
   */
  private createEdge(
    source: string,
    target: string,
    label?: string
  ): void {
    const id = `edge_${this.edgeCounter}`;
    this.edgeCounter++;

    const edge: PlaybookEdge = {
      id,
      source,
      target,
      ...(label ? { label } : {}),
    };
    this.edges.push(edge);
  }
}

/**
 * Extract metadata from markdown content (frontmatter/headers).
 */
function extractMetadata(content: string): PlaybookMetadata {
  const metadata: PlaybookMetadata = {
    title: 'Untitled Playbook',
    type: 'General',
    tooling: 'N/A',
  };

  const lines = content.split('\n');
  for (const line of lines) {
    const trimmed = line.trim();

    // Title from first H1
    if (trimmed.startsWith('# ') && metadata.title === 'Untitled Playbook') {
      metadata.title = trimmed.replace(/^#\s*/, '').trim();
    }

    // Type from **Type:** pattern
    const typeMatch = trimmed.match(/\*\*Type:\*\*\s*(.+)/);
    if (typeMatch) {
      metadata.type = typeMatch[1].trim();
    }

    // Tooling from **Tooling:** pattern
    const toolingMatch = trimmed.match(/\*\*Tooling:\*\*\s*(.+)/);
    if (toolingMatch) {
      metadata.tooling = toolingMatch[1].trim();
    }

    // Difficulty from **Difficulty:** pattern
    const difficultyMatch = trimmed.match(/\*\*Difficulty:\*\*\s*(.+)/);
    if (difficultyMatch) {
      const d = difficultyMatch[1].trim();
      if (d === 'Beginner' || d === 'Intermediate' || d === 'Advanced') {
        metadata.difficulty = d;
      }
    }

    // Estimated time
    const timeMatch = trimmed.match(/\*\*Estimated Time:\*\*\s*(.+)/);
    if (timeMatch) {
      metadata.estimatedTime = timeMatch[1].trim();
    }

    // Last updated
    const updatedMatch = trimmed.match(/\*\*Last Updated:\*\*\s*(.+)/);
    if (updatedMatch) {
      metadata.lastUpdated = updatedMatch[1].trim();
    }
  }

  return metadata;
}

/**
 * Main parse function — returns a full ParseResult including metadata.
 */
export function parseMarkdown(content: string): ParseResult {
  const start = performance.now();
  const parser = new MarkdownParser();
  const errors: string[] = [];

  let graph: PlaybookGraph;
  try {
    graph = parser.parse(content);
  } catch (err) {
    errors.push(err instanceof Error ? err.message : 'Unknown parse error');
    graph = { nodes: [], edges: [] };
  }

  const metadata = extractMetadata(content);
  const parseTimeMs = performance.now() - start;

  return { graph, metadata, errors, parseTimeMs };
}

/**
 * Quick-parse that returns just the PlaybookGraph.
 */
export function parseMarkdownToGraph(content: string): PlaybookGraph {
  const parser = new MarkdownParser();
  return parser.parse(content);
}

export default MarkdownParser;
